Port Clown, Joker, Product (Product2), Costar from bifunctors/profunctor

JordanMartinez opened the pull request:
First PR fixing https://github.com/purescript/purescript-profunctor/issues/23

I found that `Flip` couldn't be ported here because its `Functor` instance uses `lmap` which will be defined in the `Bifunctor` class.

I also named `Bifunctor.Product` `Functor.ProductBi` for a short period of time before I realized it should be called `Product2` now. Once I realized that, I renamed it to `Product2`.

rhendric commented on src/Data/Functor/Clown.purs:
```diff
@@ -0,0 +1,28 @@
+import Data.Functor.FunctorRight (class FunctorRight)
+import Data.Newtype (class Newtype)
+
+-- | Make a `Functor` over the first argument of a `Bifunctor`
```
With the reference to `` `Bifunctor` ``, this doesn't seem quite appropriate anymore. Although honestly I can't say the wording makes a lot of sense to me in its original context either.

What would people think about changing the doc comment to something like this?
```purs
-- | Lift a one-parameter type constructor to a two-parameter type constructor
-- with a phantom second parameter
```

JordanMartinez replied:
How about
> Makes a type that takes two type parameters (e.g. `Tuple`) into a `Functor` by operating on the first parameter rather than the second:
>
> ```purescript
> -- Functor instance for normal Tuple
> map show (Tuple 1 2) == (Tuple 1 "2")
> 
> -- Functor instance for Clowned Tuple
> map show (Clown (Tuple 1 2)) == Clown (Tuple "1" 2)

rhendric replied:
I don't think that's [what a Clown does](https://try.purescript.org/?gist=fb312845f43c781f7bfbb864780b1b02), though? `map`ping over a Clown has no effect on the inner value, only the phantom argument. And I don't think there needs to be special attention paid to the case when Clown is applied to a two-parameter type versus a one-parameter type, does there? A `Clown (Tuple 1 2.0)` has type `Clown (Tuple Int) Number _`, and the `Int` part doesn't have any effect on further clownery, right? (It's been a while since I've read about clowns and jokers so I could be missing something important.)

JordanMartinez replied:
Oh, whoops. You're right.

JordanMartinez replied:
How about this then?
> Lifts a type that takes one type parameter to a type that takes two type parameters where the second parameter is a phantom type.
>
> ```purescript
> data Box a = Box a
> -- these values are the same at runtime
> Box a
> Clown (Box a) :: Clown a Int
> Clown (Box a) :: Clown a String
>
> type TupleInt x = Tuple Int x
> -- these values are the same at runtime
> Tuple 4 "foo" :: TupleInt String
> Clown (Tuple 4 "foo") :: Clown (TupleInt String Boolean)
> Clown (Tuple 4 "foo") :: Clown (TupleInt String Number)
> ```

rhendric replied:
I like the general direction, although none of that actually type-checks. (The first set of examples need to be `Clown Box a Int`, etc.; the second set needs to have the parentheses around `TupleInt ...` removed and `TupleInt` itself needs to be defined without the `x` parameter so that it isn't partially applied, or it needs to be a `newtype` or something.)

rhendric commented on src/Data/Functor/Joker.purs:
```diff
@@ -0,0 +1,39 @@
+import Data.Functor.FunctorRight (class FunctorRight)
+import Data.Newtype (class Newtype, un)
+
+-- | Make a `Functor` over the second argument of a `Bifunctor`
```
Same comment as on `Clown` but with the proposed language being:

```purs
-- | Lift a one-parameter type constructor to a two-parameter type constructor
-- with a phantom first parameter
```

rhendric commented on src/Data/Functor/Product2.purs:
```diff
@@ -0,0 +1,18 @@
+import Prelude
+import Data.Functor.FunctorRight (class FunctorRight, rmap)
+
+-- | The Product of two `Bifunctor`s.
```
Another reference to `` `Bifunctor` `` which should probably change to `two-parameter type constructors` or something?

rhendric commented on src/Data/Functor/Wrap.purs:
```diff
@@ -0,0 +1,22 @@
+import Data.Functor.FunctorRight (class FunctorRight, rmap)
+import Data.Newtype (class Newtype)
+
+-- | Provides a `Functor` over the second argument of a `Bifunctor`.
```
``over the second argument of a `FunctorRight` ``?

rhendric commented on src/Data/Functor/Costar.purs:
```diff
@@ -0,0 +1,49 @@
+import Data.Functor.FunctorRight (class FunctorRight)
+import Data.Newtype (class Newtype)
+
+-- | `Costar` turns a `Functor` into a `Profunctor` "backwards".
```
This could mention that it also turns a `Contravariant` into a `Bifunctor`... but maybe that doesn't need to be said explicitly? I'm more on the fence with this one than with my other doc suggestions.

JordanMartinez commented:
@rhendric I've updated the docs. Your thoughts on them now?

rhendric reviewed (approved):
LGTM!

JordanMartinez commented:
This PR is ready for review.

kl0tl commented:
Should we specialize the second parameters of Clown, Joker, Product2, Star and Wrap to Type because of their FunctorRight instance? If yes we should then also specialize the first parameter of Costar because of the Comonad constraint in its Category instance and the first parameter of Star because of the Monad constraint in its Category instance.

JordanMartinez commented:
> Should we specialize

Probably. I've pushed the changes.

hdgarrood commented:
If you want to improve the docs for `Clown` and `Joker`, I'd recommend reading the paper ["Clowns to the Left of me, Jokers to the Right (Pearl)"](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.475.6134&rep=rep1&type=pdf), since as far as I know that is where they come from, and it's difficult to find a motivating use case for them without having read it. I think the techniques which make use of these data types should be considered advanced, so we arguably shouldn't have docs which take the position that the reader needs help understanding what `newtype Clown p a b = Clown p a` means.

hdgarrood commented:
Regarding these data types' kinds: I think it's a bit weird to specialise only the second parameter to `Type`. If the `FunctorRight` instance is a reason to specialize the second parameter to `Type`, then surely the `Bifunctor` instance is a reason to specialize the first parameter to `Type` as well? I think I prefer the idea of specializing both rather than specializing neither, and we aren't breaking anything by leaving it specialized, as they're effectively both specialized to `Type` pre-0.14.

hdgarrood commented on src/Data/Functor/Wrap.purs:
```diff
@@ -0,0 +1,27 @@
+import Data.Functor.FunctorRight (class FunctorRight, rmap)
+import Data.Newtype (class Newtype)
+
+-- | Given a type that takes two type parameters, `p`, provides a `Functor`
```
From my reading this isn't accurate; a Functor instance can only ever map over the last type parameter.

hdgarrood replied:
The real function of Wrap here is to provide a Functor instance for a data type with 2 type parameters given only a FunctorRight instance. But I can't think of a good reason why someone would provide only FunctorRight and not Functor, so I'm actually wondering if now isn't a good time to just drop Wrap entirely?

JordanMartinez replied:
Yeah, dropping it sounds good. If we ever needed to readd it, that's not a breaking change.

JordanMartinez commented:
I've removed `Star` and `Split` due to [Harry's comment](https://github.com/purescript/purescript-profunctor/issues/23#issuecomment-766330598) and removed `Wrap` due to [Harry's comment](https://github.com/purescript/purescript-functors/pull/27/files/fa2c126ddc76298db07eeea60cfe9818a682651e#diff-614d227a5974458dc73791160dc99b25df456fc1cbcc9100ba29de04cd6e6e48)

JordanMartinez commented:
> I think I prefer the idea of specializing both rather than specializing neither

Done.

JordanMartinez commented:
> "Clowns to the Left of me, Jokers to the Right (Pearl)"

I read through that yesterday. I understood the idea of refactoring two mutually recursive functions that would otherwise blow the stack into versions that would no longer blow the stack. I got lost in the parts beyond that due to all the symbols. Plus, I wasn't trying my hardest to read it either.

Perhaps the docs on `Clown` (and vice versa for `Joker`) could say
> This advance type's usage and its relation to `Joker` is best understood by reading through [Clowns to the Left, Jokers to the Right (Functional Pearl)](https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.475.6134&rep=rep1&type=pdf)

Those that want to understand them will. Those who get scared off by the paper will find a less advanced solution to their problem (or curiosity).

JordanMartinez commented:
> Should we specialize the second parameters of Clown, Joker, Product2, Star and Wrap to Type because of their FunctorRight instance? If yes we should then also specialize the first parameter of Costar

@kl0tl I also pushed these changes in this PR. The `Star` and `Wrap` one weren't pushed due to Harry's comments above.

rhendric commented on src/Data/Functor/Product2.purs:
```diff
@@ -0,0 +1,24 @@
+  show (Product2 x y) = "(Product2 " <> show x <> " " <> show y <> ")"
+
+instance functorRight :: (FunctorRight f, FunctorRight g) => FunctorRight (Product2 f g) where
+  rmap f (Product2 x y) = Product2 (rmap f x) (rmap f y)
```
If we're dropping `Wrap`, `Product2` ought to have a `Functor` instance as well, right?

JordanMartinez replied:
Ah... Yeah. I'll add that.

JordanMartinez replied:
Pushed.

hdgarrood reviewed (approved):
LGTM :tada:

JordanMartinez commented:
I've updated the docs for Clown and Joker. When I generated the docs locally, this documentation was not showing up. I'm not sure if that's because of the link or not. We can fix it later if needed. I'd like to get this in.

JordanMartinez commented:
This PR actually introduced an issue. Since it's been so long since I worked on this initially, I had forgotten a few things. Most importantly, `functors` already depends on `bifunctors`. `profunctor` does not depend on `functors` or `bifunctors`. I'm assuming `functors` should be updated to depend on `profunctor` as well? If so, then I think all of the newtypes we originally thought to port here still can be ported here (aside from those that can be only one or the other)

JordanMartinez commented:
- Profunctor bower.json file: https://github.com/purescript/purescript-profunctor/blob/master/bower.json
- Bifunctors bower.json file: https://github.com/purescript/purescript-bifunctors/blob/master/bower.json
- Functors bower.json file: https://github.com/purescript/purescript-functors/blob/master/bower.json
